### PR TITLE
PT-1026 max login counter

### PR DESCRIPTION
**Pull request by yvon-dblg**

### CODE REVIEW

File: env.sh [hunk 1]
```diff
@@ -27,6 +27,9 @@ export TIDEPOOL_SHORELINE_SERVICE='{
         "longTermDaysDuration": 30,
         "tokenDurationSecs": 2592000,
         "salt": "ADihSEI7tOQQP9xfXMO9HfRpXKu1NpIJ",
+        "maxFailedLogin": 5,
+        "delayToAllowNewLoginAttempt": 10,
```
Comment (ydrisR) [inline env.sh]:
```suggestion
        "delayBeforeNextLoginAttempt": 10,
```

File: shoreline.go [hunk 2]
```diff
@@ -55,6 +58,14 @@ const (
 func main() {
 	var config Config
 
+	// Init random number generator
+	rand.Seed(time.Now().UnixNano())
+
+	// Set some default config values
+	config.User.MaxFailedLogin = 5
+	config.User.DelayToAllowNewLoginAttempt = 10 // 10 minutes
```
Comment (ydrisR) [inline shoreline.go]:
```suggestion
	config.User.DelayBeforeNextLoginAttempt= 10 // 10 minutes
```

File: user/api.go [hunk 3]
```diff
@@ -50,11 +54,24 @@ type (
 		Salt string `json:"salt"`
 		//used for token
 		Secret string `json:"apiSecret"`
+		// Maximum number of consecutive failed login before a delay is set
+		MaxFailedLogin int `json:"maxFailedLogin"`
+		// Delay in minutes the user must wait 10min before attempting a new login if the number of
+		// consecutive failed login is more than MaxFailedLogin
+		DelayToAllowNewLoginAttempt int64 `json:"delayToAllowNewLoginAttempt"`
```
Comment (ydrisR) [inline user/api.go]:
```suggestion
		DelayBeforeNextLoginAttempt int64 `json:"delayBeforeNextLoginAttempt"`
```

File: user/api.go [hunk 4]
```diff
@@ -50,11 +54,24 @@ type (
...
 		//allows for the skipping of verification for testing
 		VerificationSecret string           `json:"verificationSecret"`
 		ClinicDemoUserID   string           `json:"clinicDemoUserId"`
 		Mailchimp          mailchimp.Config `json:"mailchimp"`
 	}
+	// LoginLimiter var needed to limit the max login attempt on an account
+	LoginLimiter struct {
+		mutex               sync.Mutex
+		userNamesInProgress *list.List
+		nInProgress         int
```
Comment (ydrisR) [inline user/api.go]:
```suggestion
		totalInProgress         int
```

File: user/api.go [hunk 5]
```diff
@@ -50,11 +54,24 @@ type (
...
+		MaxConcurrentLogin int `json:"maxConcurrentLogin"`
 		//allows for the skipping of verification for testing
 		VerificationSecret string           `json:"verificationSecret"`
 		ClinicDemoUserID   string           `json:"clinicDemoUserId"`
 		Mailchimp          mailchimp.Config `json:"mailchimp"`
 	}
+	// LoginLimiter var needed to limit the max login attempt on an account
+	LoginLimiter struct {
+		mutex               sync.Mutex
+		userNamesInProgress *list.List
```
Comment (ydrisR) [inline user/api.go]:
```suggestion
		usersInProgress *list.List
```

File: user/api.go [hunk 6]
```diff
@@ -112,13 +128,17 @@ func InitApi(cfg ApiConfig, store Storage, metrics highwater.Client) *Api {
...
-	return &Api{
+	api := Api{
 		Store:            store,
 		ApiConfig:        cfg,
 		metrics:          metrics,
 		logger:           logger,
 		mailchimpManager: mailchimpManager,
 	}
+
+	api.loginLimiter.userNamesInProgress = list.New()
```
Comment (ydrisR) [inline user/api.go]:
```suggestion
	api.loginLimiter.usersInProgress = list.New()
```

File: user/api.go [hunk 7]
```diff
@@ -266,7 +288,7 @@ func (a *Api) CreateUser(res http.ResponseWriter, req *http.Request) {
 		a.sendError(res, http.StatusInternalServerError, STATUS_ERR_CREATING_USR, err)
 
 	} else if len(existingUser) != 0 {
-		a.sendError(res, http.StatusConflict, STATUS_USR_ALREADY_EXISTS)
+		a.sendError(res, http.StatusInternalServerError, STATUS_ERR_CREATING_USR, fmt.Sprintf("User '%s' already exists", *newUserDetails.Username))
```
Comment (ydrisR) [inline user/api.go]: Why changing this behaviour?

File: user/api.go [hunk 8]
```diff
@@ -578,29 +600,47 @@ func (a *Api) DeleteUser(res http.ResponseWriter, req *http.Request, vars map[st
...
+	time.Sleep(time.Millisecond * time.Duration(rand.Int63n(100)))
+
+	code, elem := a.appendUserLoginInProgress(user)
+	defer a.removeUserLoginInProgress(elem)
+	if code != http.StatusOK {
+		a.sendError(res, http.StatusUnauthorized, STATUS_NO_MATCH, fmt.Sprintf("User '%s' has too many ongoing login: %d", user.Username, a.loginLimiter.nInProgress))
 
 	} else if results, err := a.Store.FindUsers(user); err != nil {
-		a.sendError(res, http.StatusInternalServerError, STATUS_ERR_FINDING_USR, err)
+		a.sendError(res, http.StatusUnauthorized, STATUS_NO_MATCH, STATUS_USER_NOT_FOUND, err)
```
Comment (ydrisR) [inline user/api.go]: why changing the returned code here? it is indeed an internal error if the db query fails...

File: user/api.go [hunk 9]
```diff
@@ -578,29 +600,47 @@ func (a *Api) DeleteUser(res http.ResponseWriter, req *http.Request, vars map[st
...
 
 	} else if len(results) != 1 {
-		a.sendError(res, http.StatusUnauthorized, STATUS_NO_MATCH, fmt.Sprintf("Found %d users matching %#v", len(results), user))
+		a.sendError(res, http.StatusUnauthorized, STATUS_NO_MATCH, fmt.Sprintf("User '%s' have %d matching results", user.Username, len(results)))
```
Comment (ydrisR) [inline user/api.go:627]: Really need to change this log message?

File: user/api.go [hunk 10]
```diff
@@ -966,3 +1010,26 @@ func (a *Api) removeUserPermissions(groupId string, removePermissions clients.Pe
...
+
+// UpdateUserAfterFailedLogin update the user failed login infos in database
+func (a *Api) UpdateUserAfterFailedLogin(u *User) error {
+	if u.FailedLogin == nil {
+		u.FailedLogin = new(FailedLoginInfos)
+	}
+	u.FailedLogin.Count++
+	u.FailedLogin.Total++
+	if u.FailedLogin.Count >= a.ApiConfig.MaxFailedLogin {
+		nextAttemptTime := time.Now().Add(time.Minute * time.Duration(a.ApiConfig.DelayToAllowNewLoginAttempt))
```
Comment (ydrisR) [inline user/api.go]:
```suggestion
		nextAttemptTime := time.Now().Add(time.Minute * time.Duration(a.ApiConfig.DelayBeforeNextLoginAttempt))
```

File: user/api_test.go [hunk 11]
```diff
@@ -38,12 +39,15 @@ var (
...
-		LongTermKey:        "thelongtermkey",
-		Salt:               "a mineral substance composed primarily of sodium chloride",
-		VerificationSecret: "",
-		ClinicDemoUserID:   "00000000",
+		Secret:                      "This is a local API secret for everyone. BsscSHqSHiwrBMJsEGqbvXiuIUPAjQXU",
+		TokenDurationSecs:           TOKEN_DURATION,
+		LongTermKey:                 "thelongtermkey",
+		Salt:                        "a mineral substance composed primarily of sodium chloride",
+		MaxFailedLogin:              5,
+		DelayToAllowNewLoginAttempt: 10,
```
Comment (ydrisR) [inline user/api_test.go]:
```suggestion
		DelayBeforeNextLoginAttempt: 10,
```

File: user/api_test.go [hunk 12]
```diff
@@ -1360,6 +1365,28 @@ func Test_Login_Error_NoPassword(t *testing.T) {
...
+
+	headers := http.Header{}
+	headers.Add("Authorization", authorization)
+	response := T_PerformRequestHeaders(t, "POST", "/login", headers)
+	T_ExpectErrorResponse(t, response, 401, "No user matched the given details")
+}
+
+func Test_Login_Error_AccountLock(t *testing.T) {
+	authorization := T_CreateAuthorization(t, "a@b.co", "password")
+	nextAttemptTime := time.Now().Add(time.Minute * time.Duration(FAKE_CONFIG.DelayToAllowNewLoginAttempt))
```
Comment (ydrisR) [inline user/api_test.go]:
```suggestion
	nextAttemptTime := time.Now().Add(time.Minute * time.Duration(FAKE_CONFIG.DelayBeforeNextLoginAttempt))
```

File: user/helpers.go [hunk 13]
```diff
@@ -142,3 +143,34 @@ func (a *Api) asSerializableUser(user *User, isServerRequest bool) interface{} {
 	}
 	return serializable
 }
+
+func (a *Api) appendUserLoginInProgress(user *User) (code int, elem *list.Element) {
+	a.loginLimiter.mutex.Lock()
+	defer a.loginLimiter.mutex.Unlock()
+
+	// Simple rate limiter
+	a.loginLimiter.nInProgress++
```
Comment (ydrisR) [inline user/helpers.go]:
```suggestion
	a.loginLimiter.totalInProgress++
```

File: user/helpers.go [hunk 14]
```diff
@@ -142,3 +143,34 @@ func (a *Api) asSerializableUser(user *User, isServerRequest bool) interface{} {
...
+	if a.loginLimiter.nInProgress > a.ApiConfig.MaxConcurrentLogin {
```
Comment (ydrisR) [inline user/helpers.go]:
```suggestion
	if a.loginLimiter.totalInProgress > a.ApiConfig.MaxConcurrentLogin {
```

**Review by ydrisR — CHANGES_REQUESTED** [review]
Nice work globally :)
I propose to change a couple of variable names. 

One problematic changed (at least from my point of view) if the changes on the HTTP return codes. I don't think there is a need to change the current behaviour.
Also we should be carefull about what is logged by the service: should we log user names and other personal infos in the log files? Probably something to discuss as a team.

File: user/api.go (hunk 9 above)
Comment (yvon-dblg) [reply user/api.go:627]: It was to have all logged message on the same pattern.
Should help @CedricPerotto with ELK.

File: user/api.go (hunk 7 above)
Comment (yvon-dblg) [reply user/api.go]: OWASP recommandations, see https://owasp.org/www-project-cheat-sheets/cheatsheets/Authentication_Cheat_Sheet.html

File: user/api.go (hunk 8 above)
Comment (yvon-dblg) [reply user/api.go]: OWASP recommandations, see https://owasp.org/www-project-cheat-sheets/cheatsheets/Authentication_Cheat_Sheet.html

File: user/api.go (hunk 7 above)
Comment (ydrisR) [reply user/api.go]: I would rather return 403 Forbiden or 401 unauthorized

**Review by ydrisR — APPROVED** [review]
LGTM

**Pull request merged**